cli/cppcheckexecutor.cpp

```diff
@@ -383,15 +382,18 @@ static void CppcheckSignalHandler(int signo, siginfo_t * info, void * context)
 {
     int type = -1;
     pid_t killid;
-#if defined(__linux__) && defined(REG_ERR)
```

Comment:
this is not just a refactoring, it changes the structure a bit. If somebody will read this later he needs to consider what will happen in the "__linux__ is not defined and REG_ERR is defined" case that we explicitly try to handle. 

It's better that the focus is "proper and readable" code than "make iwyu happy".

Reply:
Actually this has nothing to do with iwyu. If you have `__linux__` defined but `REG_ERR` is not you will be using the PID since you didn't get the thread ID but the process ID. It's two different features which are checked for inclusively in a single check.

Reply:
Is `type = (int)uc->uc_mcontext.gregs[REG_ERR] & 2;` valid code when the `__linux__` is not defined? Would it be more proper like:

    #ifdef __linux__
        const pid_t killid = (pid_t) syscall(SYS_gettid);
    #ifdef REG_ERR
        const ucontext_t* const uc = reinterpret_cast<const ucontext_t*>(context);
        const int type = uc ? (int)uc->uc_mcontext.gregs[REG_ERR] & 2 : -1;
    #endif
    #else
          UNUSED(context);
          const int type = -1;
    #endif

?

That code is also more cludgy than the original code but is that better?

Reply:
I can't really tell without knowing why `REG_ERR` is optional in the first place.

`__linux__` is actually not the proper way to check for this - it should actually be done as suggested in the manpages - see https://man7.org/linux/man-pages/man2/syscall.2.html and https://man7.org/linux/man-pages/man7/feature_test_macros.7.html. Unfortunately that might not work since some implementations and platforms do not abide to that (looking at you `musl`).

`syscall()` is definitely Linux-specific but that register stuff seems to be *nix stuff so might also available on other systems.

I would personally get rid of the conditional `REG_ERR` handling and flush out where it is not available so we can properly document it. Looking at the code I think we could get completely rid of it since I see no point in having that detailed output of `SIGSEGV`. I don't think it will be of any help if this ever occurs since we would require a stacktrace after all and most likely we have a sample to reproduce it so we will debug it anyways or use things like valgrind or sanitizers to get the required data.

Reply:
Or you could just look at the commit log - https://github.com/danmar/cppcheck/commit/9438b49bfbcf3b060bf318fbc5e63dc1da72650a

```
Fix build on PowerPC

PowerPC does not provide ucontext.gregs or the REG_ERR define.
```